File: app/src/main/java/com/example/notebook/NoteEntity.java
package com.example.notebook;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Calendar;
import java.util.Date;
import java.util.UUID;

public class NoteEntity implements Parcelable {
    public static final Creator<NoteEntity> CREATOR = new Creator<NoteEntity>() {
        @Override
        public NoteEntity createFromParcel(Parcel in) {
            return new NoteEntity(in);
        }

        @Override
        public NoteEntity[] newArray(int size) {
            return new NoteEntity[size];
        }
    };
    private final String id;
    private String title;
    private String description;
    private Date date;

    NoteEntity(String id,
               String title,
               String description,
               Date date) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.date = date;
    }

    protected NoteEntity(Parcel in) {
        id = in.readString();
        title = in.readString();
        description = in.readString();
    }

    public static String generateNewId() {
        return UUID.randomUUID().toString();
    }

    public static Date getCurrentDate() {
        return Calendar.getInstance().getTime();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(title);
        dest.writeString(description);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public Date getDate() {
        return date;
    }

    public void update(NoteEntity newNote) {
        description = newNote.description;
        title = newNote.title;
    }
}

File: app/src/main/java/com/example/notebook/NoteEntity.kt
package com.example.notebook

import android.os.Parcelable
import android.os.Parcel
import android.os.Parcelable.Creator
import java.util.*

class NoteEntity : Parcelable {
    val id: String?
    var title: String?
        private set
    var description: String?
        private set
    var date: Date? = null
        private set

    internal constructor(
        id: String?,
        title: String?,
        description: String?,
        date: Date?
    ) {
        this.id = id
        this.title = title
        this.description = description
        this.date = date
    }

    protected constructor(`in`: Parcel) {
        id = `in`.readString()
        title = `in`.readString()
        description = `in`.readString()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(title)
        dest.writeString(description)
    }

    override fun describeContents(): Int {
        return 0
    }

    fun update(newNote: NoteEntity) {
        description = newNote.description
        title = newNote.title
    }

    companion object {
        @JvmField
        val CREATOR: Creator<NoteEntity?> = object : Creator<NoteEntity?> {
            override fun createFromParcel(`in`: Parcel): NoteEntity? {
                return NoteEntity(`in`)
            }

            override fun newArray(size: Int): Array<NoteEntity?> {
                return arrayOfNulls(size)
            }
        }

        @JvmStatic
        fun generateNewId(): String {
            return UUID.randomUUID().toString()
        }

        @JvmStatic
        val currentDate: Date
            get() = Calendar.getInstance().time
    }
}